src/gromacs/nblib/gmxsetup.h

```diff
@@ -64,24 +64,35 @@
 namespace nblib
 {
 
-struct NbvSetupUtil
+class NbvSetupUtil
 {
+public:
     NbvSetupUtil(SimulationState system, const NBKernelOptions& options);
 
-    void unpackTopologyToGmx();
-
-    std::unique_ptr<nonbonded_verlet_t> setupNbnxmInstance();
-
+    //! Sets up and returns a GmxForceCalculator
     std::unique_ptr<GmxForceCalculator> setupGmxForceCalculator();
 
-    SimulationState                  system_;
-    std::shared_ptr<NBKernelOptions> options_;
-
     //! Storage for parameters for short range interactions.
     std::vector<real> nonbondedParameters_;
 
     //! Particle info where all particles are marked to have Van der Waals interactions
     std::vector<int> particleInfoAllVdw_;
```

Comment:
nonbondedParameters_ and particleInfoAllVdw_ is internal state and needs to be private.
This class still has none of the functions in the public interface that we discussed about, i.e. a function to add
the nonbonded parameters, the atomProperties (VdW, chargest, etc), the AtomTypes etc.
You don't want to pass the entire system and then set all that's needed in an opaque way internally. The necessary data should be fed in piece by piece through a public interface, since this class is supposed to be a builder for GmxForceCalculator.

Reply:
I absolutely get this and that would be the goal of the builder PR. This PR is more to consolidate things within `GmxSetupUtils` by creating functions to carry specific arguments. The members are now private. The public interface would be expanded and used in the forthcoming PR.

Reply:
Important design discussion. I get the value of having those as separate functions to be called with specific arguments. And that is exactly how this is shaping up.

However, having code being called in the public interface is probably not necessary. We can have one builder super-function that does the step-by-step building which is essential for debugging, but having those calls live inside `ForceCalculator` is something whose benefits I don't see yet.

```
ForceCalculator::ForceCalculator(const SimulationState& system, const NBKernelOptions& options)
{
    NbvSetupUtil nbvSetupUtil(system, options);

    gmxForceCalculator_ = nbvSetupUtil.setupGmxForceCalculator();
}
```

The above snippet looks super clean to work with. The functions you described can be called inside the `setupGmxForceCalculator`

Reply:
@ejjordan Comments/criticism/guidance requested here.

Reply:
From the user's perspective, it will not matter how complex or not the builder for GmxForceCalculator is, since it is not a user interactable object. What we want is an interface for the GmxForceCalculatorBuilder that has a similar feel to the TopologyBuilder. What that means is that we want to be able to write a script that one by one adds the pieces to the object that is being built. Something like the following.

`GmxForceCalculatorBuilder fcb;`
`fcb.addMDAtoms(...);`
`fcb.addNonBondedParameters(...);`
`fcb.addInteractionConst(...);`
`// Add more stuff`
`GmxForceCalculator fc = fcb.build();`

Reply:
I am trying to understand the benefits of making it feel like `TopologyBuilder`.

The `GmxForceCalculator` uses all the data in `SimulationState`. and `NBKernelOptions` even if the `NbvSetupUtil` doesn't. For this reason, the builder must pass both these objects down.

The pieces for this are not scattered the way the data for `Topology` is. The latter needs a variety of molecules constructed in scope.

Reply:
That analogy is not entirely valid in the case `GmxForceCalculator` as there isn't much to its own internal state. There is however a multitude of steps in constructing `GmxForceCalculator` and its internal members as you can see in `setupGmxForceCalculator()`.  You can see a semblance of such a script here.

I do not see benefits yet in making those calls in the constructor of `ForceCalculator`. Please clarify this.

Reply:
@ejjordan Exactly! @kanduri Yes, the analogy to TopologyBuilder is valid and of course the calculator builder has a lot of internal state, namely all the data that Joe mentioned above.
What Joe outlined above is exactly what we decided to do and I am very confused why this is still unclear or at least, in case you have a different vision, you didn't voice it two weeks ago when we were all in the same room.

Reply:
@kanduri, Yes! What @ejjordan and @sebkelle1 are saying is what we agreed in the hackathon.

Reply:
Btw what you outlined is the director code. It can be it's own function that's called in the ForceCalculator, but it does neither belong into the gmx calculator builder nor into the gmx calculator itself.

Reply:
@sebkelle1 @ejjordan 
`NbvSetupUtl` has just two internal members namely `nonbondedParameters_` and `particleInfoAllVdw_` which basically come from `NBKernelOptions`. Unlike the `TopologyBuilder` which takes new molecules, and updates all their internal members based on these inputs. The references to the other two are just stored as they come and no logic.

Because there are just two members that basically need the same input, this part is handled within the constructor using dedicated functions so that this process is clearly separated. 

Construction of `GmxForceCalculator` is a multi-stage one and a script for that exists in `setupGmxForceCalculator()`. Since this is meant to be a builder, it certainly belongs there. Each member inside `GmxForceCalculator` is filled individually using dedicated functions wherever possible.

I can do what you asked for if you all really want it, but I don't see the benefits yet of pushing this script inside the constructor of `ForceCalculator`.

Reply:
The script does not live in `ForceCalculator` but in an Impl class that is called there.

Reply:
So do we want another Impl object outside `NbvSetupUtil` that orchestrates the filling of each member of `GmxForceCalculator`?

I am not seeing benefits of doing so yet.

Reply:
There are a number of benefits.
1) What is need for object construction is clear since it all happens in one place.
2) As we add functionality that requires more steps in the builder, it is obvious where to add those steps.
3) Steps of building that have lots of logic can drive refactoring efforts of gromacs internals to make the interface layer thinner.
4) Dependencies between steps can be easily managed with asserts in the steps. This can also focus refactoring efforts since it helps show data flow between objects.

Basically having this layer helps keep complexity manageable through expansion of scope and refactoring.

Reply:
Please check the new push. There is not a `GmxBuildDirector` class introduced.

1. With segregated functions for each field of the `GmxForceCalculator`, that clarification had been introduced.

2. There is a specific function to add those steps both before and now. 

3. Agreed. So having such a script is essential, the discussion was on where that should live. 

4. Good idea. I am drafting a small set of design goals which @sebkelle1 and I spoke about just a while ago.

Reply:
Also, renaming classes would be done after both of the outstanding PRs are merged. Just want to prevent conflicts that emerge solely due to the rename.